src/vivarium/framework/population/population_view.py

```diff
@@ -208,46 +205,143 @@ def update(self, population_update: Union[pd.DataFrame, pd.Series]) -> None:
+            4. The update columns are a subset of the columns managed by this
+               PopulationView.
+            5. The update index is a subset of the existing state table index.
+               PopulationViews don't make rows, they just fill them in.
+
+        For initial population creation additional preconditions are documented in
+        :meth:`PopulationView._ensure_coherent_initialization`. Outside population
+        initialization, we require that all columns in the update to be present in
+        the existing state table. When new simulants are added in the middle of the
+        simulation, we require that only one component provide updates to a column.
+
+        Parameters
+        ----------
+        population_update
+            The update to the simulation state table.
+        state_table
+            The existing simulation state table.
+        view_columns
+            The columns managed by this PopulationView.
+        creating_initial_population
+            Whether the initial population is being created.
+        adding_simulants
+            Whether new simulants are currently being initialized.
+
+        Returns
+        -------
+        pandas.DataFrame
+            The input data formatted as a DataFrame.
+
+        Raises
+        ------
+        TypeError
+            If the population update is not a :class:`pandas.Series` or a
+            :class:`pandas.DataFrame`.
+        PopulationError
+            If the update violates any preconditions relevant to the context in which
+            the update is provided (initial population creation, population creation on
+            time steps, or population state changes on time steps).
+
+        """
+        assert not (creating_initial_population and not adding_simulants)
+
+        population_update = PopulationView._coerce_to_dataframe(
+            population_update,
+            view_columns,
+        )
+
+        unknown_simulants = population_update.index.difference(state_table.index)
+        if not unknown_simulants.empty:
+            raise PopulationError(
+                "Population updates must have an index that is a subset of the current "
+                f"population state table. {len(unknown_simulants)} simulants were provided "
+                f"in an update with no matching index in the existing table."
+            )
+
+        if creating_initial_population:
+            PopulationView._ensure_coherent_initialization(population_update, state_table)
+        else:
+            new_columns = list(set(population_update).difference(state_table))
+            if new_columns:
+                raise PopulationError(
+                    f"Attempting to add new columns {new_columns} to the state table "
+                    f"outside the initial population creation phase."
+                )
+
+            if adding_simulants:
+                for column in population_update:
```

Comment:
It would be better to loop through and get all columns where the conditional below holds, and write out an error message with all of them, so you can get them all at once if you have multiple, rather than getting them one at a time and having to re-run to learn about the next one.

Reply:
Done, but this helps less than you think it would as the error is triggered by one call from one component (which means we can still get it multiple times)